server/environment/manager.go

```diff
@@ -0,0 +1,330 @@
+}
+
+func (r *Repository) Get(ctx context.Context, id id.ID) (Environment, error) {
+	stmt, err := r.db.Prepare(getQuery + " WHERE e.id = $1")
+
+	if err != nil {
+		return Environment{}, fmt.Errorf("prepare: %w", err)
+	}
+	defer stmt.Close()
+
+	environment, err := r.readEnvironmentRow(ctx, stmt.QueryRowContext(ctx, id))
+	if err != nil {
+		return Environment{}, err
+	}
+
+	return environment, nil
+}
+
+func (r *Repository) EnvironmentIDExists(ctx context.Context, id string) (bool, error) {
+	exists := false
+
+	row := r.db.QueryRowContext(ctx, idExistsQuery, id)
+
+	err := row.Scan(&exists)
+
+	return exists, err
+}
+
+func (r *Repository) readEnvironmentRow(ctx context.Context, row scanner) (Environment, error) {
+	environment := Environment{}
+
+	var (
+		jsonValues []byte
+	)
+	err := row.Scan(
+		&environment.ID,
+		&environment.Name,
+		&environment.Description,
+		&environment.CreatedAt,
+		&jsonValues,
+	)
+
+	switch err {
+	case sql.ErrNoRows:
+		return Environment{}, err
+	case nil:
+		err = json.Unmarshal(jsonValues, &environment.Values)
+		if err != nil {
+			return Environment{}, fmt.Errorf("cannot parse environment: %w", err)
+		}
+
+		return environment, nil
+	default:
+		return Environment{}, err
+	}
+}
+
+func (r *Repository) countEnvironments(ctx context.Context, query string) (int, error) {
+	var (
+		count  int
+		params []any
+	)
+
+	sql := countQuery + query
+
+	err := r.db.
+		QueryRowContext(ctx, sql, params...).
+		Scan(&count)
+
+	if err != nil {
+		return 0, err
+	}
+	return count, nil
+}
+
+func (r *Repository) insertIntoEnvironments(ctx context.Context, environment Environment) (Environment, error) {
+	stmt, err := r.db.Prepare(insertQuery)
+	if err != nil {
+		return Environment{}, fmt.Errorf("sql prepare: %w", err)
+	}
+	defer stmt.Close()
+
+	jsonValues, err := json.Marshal(environment.Values)
+	if err != nil {
+		return Environment{}, fmt.Errorf("encoding error: %w", err)
+	}
+
+	_, err = stmt.ExecContext(
+		ctx,
+		environment.ID,
+		environment.Name,
+		environment.Description,
+		environment.CreatedAt,
+		jsonValues,
+	)
+
+	if err != nil {
+		return Environment{}, fmt.Errorf("sql exec: %w", err)
+	}
+
+	return environment, nil
+}
+
+func (r *Repository) updateIntoEnvironments(ctx context.Context, environment Environment, oldId id.ID) (Environment, error) {
+	stmt, err := r.db.Prepare(updateQuery)
+	if err != nil {
+		return Environment{}, fmt.Errorf("sql prepare: %w", err)
+	}
+	defer stmt.Close()
+
+	jsonValues, err := json.Marshal(environment.Values)
+	if err != nil {
+		return Environment{}, fmt.Errorf("encoding error: %w", err)
+	}
+
+	_, err = stmt.ExecContext(
+		ctx,
+		oldId,
+		environment.Slug(),
+		environment.Name,
+		environment.Description,
+		environment.CreatedAt,
+		jsonValues,
+	)
+
+	if err != nil {
+		return Environment{}, fmt.Errorf("sql exec: %w", err)
+	}
+
+	return environment, nil
+}
+
+func (r *Repository) Provision(ctx context.Context, environment Environment) error {
```

Comment:
With this approach, we can only provision one environment, right?

Reply:
No. Each manager provision only one resource at a time. But the provisioner is responsible for reading all resources from the provision YAML file and iterating through them and delegating the provision to this function. So, if there are three environments to be provisioned, this function will be called three times.